paper.md

```diff
@@ -0,0 +1,105 @@
+
+# Related work
+
+To address the performance issues in Qiskit, we explored several graph library alternatives. _SNAP_ [@Leskovec2016], _graphtool_ [@Peixoto2014], and _igraph_ [@Csardi2006] are stable Python libraries written in C or C++ that can replace _NetworkX_.
+
+However, there was a strong desire to keep the flexibility that _NetworkX_ provided for exploring and interacting with the graphs, which precluded custom data structures. The investigated graph libraries either had issues integrating with Qiskit or APIs that were too rigid, such that the migration of existing code was more complex than desired. Thus, the main contribution of _retworkx_ is keeping the ease of use of _NetworkX_ without sacrificing performance.
+
+# Graph data structures
+
+_retworkx_ provides two core data structures: `PyGraph` and `PyDiGraph`. They correspond to undirected and directed graphs, respectively. Graphs describe a set of nodes and the edges connecting pairs of those nodes. Internally, _retworkx_ leverages the _petgraph_ [@bluss2021] and the _PyO3_ libraries [@Hewitt2021] to store the graphs.
+
+Nodes and edges of the graph may also be associated with weights. Weights can contain arbitrary data, such as node labels or edge lengths. Any Python object can be a weight, which makes the library flexible because no assumptions are made about the weight types. 
+
+A defining characteristic of _retworkx_ graphs is that each node maps to a non-negative integer node index, and similarly, each edge maps to an edge index. Those indices uniquely determine nodes and edges during the graph object's lifetime. Moreover, the indices provide a clear separation between the underlying graph structure and the data associated with weights.
+
+_retworkx_ operates on weights with callbacks. Callbacks are functions that take weights and return statically typed data. They resemble the named attributes in _NetworkX_. Callbacks are beneficial because they bridge the arbitrary stored data with the static types _retworkx_ expects. We illustrate indices and callbacks usage with an example.
+
+## Example
+
+```python
+import retworkx
+
+graph = retworkx.PyGraph()
+
+a = graph.add_node("A")
+b = graph.add_node("B")
+c = graph.add_node("C")
+
+graph.add_edges_from([(a, b, 1.5), (a, c, 5.0), (b, c, 2.5)])
+retworkx.dijkstra_shortest_paths(graph, a, c, weight_fn=float)
+```
+
+In the provided example, we find the shortest path from $A$ to $C$ using Dijkstra's algorithm. Firstly, we create the graph and add three nodes labeled A, B, and C. Each time **`add_node`** is called, the method returns the node index associated with the newly added node.
```

Comment:
Do we need a citation for `Dijkstra's algorithm` here? (I know it's famous and well known but just seems like something missing)

Reply:
```suggestion
In the provided example, we find the shortest path from $A$ to $C$ using Dijkstra's algorithm [@Dijkstra1959ANO]. Firstly, we create the graph and add three nodes labeled A, B, and C. Each time **`add_node`** is called, the method returns the node index associated with the newly added node.
```

Reply:
Dijkstra's algorithm is assumed to be common knowledge, but oh well. For example, I cited Rust because it is fairly recent (< 10 years). I will read the 1959 paper before citing it

Reply:
I am including it, the paper is relevant